src/coreclr/jit/codegenxarch.cpp

```diff
@@ -4397,6 +4398,37 @@ void CodeGen::genCodeForShift(GenTree* tree)
             inst_RV_SH(ins, size, tree->GetRegNum(), shiftByValue);
         }
     }
+#if defined(TARGET_64BIT)
+    else if (compiler->compOpportunisticallyDependsOn(InstructionSet_BMI2) && tree->OperIs(GT_LSH, GT_RSH, GT_RSZ) &&
+             (genActualType(targetType) == TYP_LONG))
+    {
+        // Try to emit shlx, sarx, shrx if BMI2 is available instead of mov+shl, mov+sar, mov+shr.
+        switch (tree->OperGet())
+        {
+            case GT_LSH:
+                ins = INS_shlx;
+                break;
+
+            case GT_RSH:
+                ins = INS_sarx;
+                break;
+
+            case GT_RSZ:
+                ins = INS_shrx;
+                break;
+
+            default:
+                unreached();
+        }
+
+        regNumber shiftByReg = shiftBy->GetRegNum();
+        emitAttr  size       = emitTypeSize(tree);
+        GetEmitter()->emitIns_R_R_R(ins, size, tree->GetRegNum(), shiftByReg, operandReg);
```

Comment:
Might be worth a note that we don't currently support the contained form due to more complex changes being needed in the emitter.

Ideally we'd fix up the emitter and use `inst_RV_RV_TT` instead so that we can emit `shlx r32a, r/m32, r32b`. Someone would need to walk through the relevant `IF_RWR_RRD_*RD` formats and ensure that it's all handled correctly (noting that technically the format is `IF_RWR_*RD_RRD` but that should be the same as `IF_RWR_RRD_*RD` with swapping op1/op2, like we do for a couple other BMI2 instructions).

Reply:
Opened https://github.com/dotnet/runtime/issues/67314.

Reply:
As @tannergooding , lets have a comment about containment and also specify that here, the operands are swapped because the way operands are encoded in these 3 instructions.

Reply:
Added commnets

> As @tannergooding , lets have a comment about containment and also specify that here, the operands are swapped because the way operands are encoded in these 3 instructions.